### PR TITLE
Add a namespace-glyph when spell-checking+add-using.  

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
this helps cal out that this is going to do more than just fixup the name, it's also going to add an import/using

Also fix an issue where we didn't have't glyph when adding a reference through 'add using'.  We used to have this, but it was lost as part of a large refactoring in this space.

When spell-checking and adding a using, we now look like:

image

When adding a reference, things have been restored so we look like:

image

Or if it's a project reference:

image

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @jcouv .

**Comment by justcla** [comment]
I don't like this. If it's a lightbulb fix to an issue on the line, there should be a lightbulb icon beside it so that it's clear this is a suggestion being offered to address an issue on the line.

**Comment by CyrusNajmabadi** [comment]
> I don't like this. If it's a lightbulb fix to an issue on the line, there should be a lightbulb icon beside it so that it's clear this is a suggestion being offered to address an issue on the line.

I think we will need the facility for individual fixes to override what default we may decide goes along with an item.

Note: this is functionality we already have today and we already expose.  Analyzers can, and do, provide Glyphs for items.  So, even if we add automatic Glyphs, we shoudl not be overriding what the analyzer wants.

These glyphs serve important purposes, and we should not reduce signal and increase noise.

--

Now, if we wanted to have some other way of indicating the lightbulb (perhaps with a glyph on the right, or some other sort of adornment), i would be ok with htat.  But the fallback glyphs should not override the analyzer glyphs.

**Comment by justcla** [comment]
> we should not be overriding what the analyzer wants.
These glyphs serve important purposes, and we should not reduce signal and increase noise.

I can't see solid justification for this statement. If we find a better way to do this, then we should do things a better way.
In the case of a lightbulb menu, it could be argued that it's more important to clearly group fixes and tools than to have a range of icons that indicate the package being used.
And I will have been arguing that point. :-)

**Comment by CyrusNajmabadi** [comment]
> In the case of a lightbulb menu, it could be argued that it's more important to clearly group fixes and tools

That's a very interesting idea.  Instead of individual items having glyphs.  What if we could just have sections.  i.e.: here's the fix section (with a glyph).  Here's the tools section (with a glyph).

Then you have the "grouping" you're talking about, but you don't need to repeat the same Glyph N times throughout the list.  It also means individual items can specify their glyph.

>  than to have a range of icons that indicate the package being used.

"it could be argued"... true.  But i would argue the individual glyph is more important then the 'broad category' glyph.

In the above cases, all of the issues 'fix' the problem.  I know that because i'm aware of the code i wrote, and the item strongly conveys what it will do.  So, for example "i wrote an unbound type".  "ah, i can fix this by generating a type.  or changing the name (**oh** and adding a using).  or adding this reference".  In call cases, i'm fixing the issue, so i don't really need that explained again***.  However, i do benefit from realizing that this fix goes beyond what i might have expected.

*** Note this is not true in all cases.  In some cases today it's unclear waht's actually a fix and what isn't.  I do think we should improve that to make it clearer.  But it's not universal that this can only be conveyed by having each item have it's own glyph.

**Comment by Neme12** [comment]
No tests?

**Comment by CyrusNajmabadi** [comment]
I'm working on extending the code-action testing harness to specify expected glyphs.  But it's a lot of updating.

**Comment by Neme12** [comment]
How so? How is it different to this:
http://source.roslyn.io/#Roslyn.Services.Editor.CSharp.UnitTests/Completion/CompletionProviders/DeclarationNameCompletionProviderTests.cs,199

**Comment by Neme12** [comment]
Never mind, that is completion... I am stupid

**Comment by CyrusNajmabadi** [comment]
> How so? How is it different to this:

Because those are completoin tests.  I'm in analyzers/lightbulbs.

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by Neme12** [comment]
Another code action that alters the project and could potentially use an icon to help draw attention to it would be "upgrade project". Do you think there's an icon that would be appropriate for that?

**Comment by CyrusNajmabadi** [comment]
test this please

**Comment by CyrusNajmabadi** [comment]
@mavasani @jcouv this is available for review.

**Comment by CyrusNajmabadi** [comment]
> Another code action that alters the project (or even the whole solution) and could potentially use an icon to help draw attention to it would be "upgrade project". Do you think there's an icon that would be appropriate for that?

I think we could just show the appropriate project (i.e. CSharpProject or VBProject) icon for that case.

**Comment by CyrusNajmabadi** [comment]
@mavasani @jcouv @dotnet/roslyn-ide can i get some reviews here?  Note this fixes regressions in our previous behavior.  i've added tests to catch these regressions.

**Comment by CyrusNajmabadi** [comment]
@mavasani @jcouv @dotnet/roslyn-ide can i get some reviews here? Note this fixes regressions in our previous behavior. i've added tests to catch these regressions.

**Comment by CyrusNajmabadi** [comment]
test windows_debug_vs-integration_prtest

**Comment by CyrusNajmabadi** [comment]
test ubuntu_16_mono_debug_prtest

**Comment by mavasani** [comment]
@dotnet-bot retest ubuntu_16_mono_debug_prtest

**Comment by mavasani** [comment]
@dotnet/roslyn-ide @jcouv for one more review.

**Comment by CyrusNajmabadi** [comment]
@jcouv @dotnet/roslyn-ide @jasonmalinowski Can i get some eyes on this?  It's been 3 weeks so far.  Thanks!

**Comment by CyrusNajmabadi** [comment]
@jcouv @dotnet/roslyn-ide @jasonmalinowski Can i get some eyes on this? It's been 3 weeks so far. Thanks!

**Comment by CyrusNajmabadi** [comment]
@jinujoseph @jcouv Doesn't feel comfortable reviewing this.  Who is a suitable person who can look at this?  Thanks!

(note: it's been 3+ weeks now waiting on this.  i'd really like this to go in soon.  Thanks!)

**Comment by jinujoseph** [comment]
@heejaechang  for second review

File: src/Features/Core/Portable/AddImport/References/ProjectSymbolReference.cs [hunk 1]
```diff
@@ -36,7 +36,11 @@ protected override ImmutableArray<string> GetTags(Document document)
             {
                 return document.Project.Id == _project.Id
                     ? ImmutableArray<string>.Empty
-                    : WellKnownTagArrays.AddReference;
+                    : _project.Language == LanguageNames.CSharp
+                        ? WellKnownTagArrays.CSharpProject
```
Comment (heejaechang) [inline src/Features/Core/Portable/AddImport/References/ProjectSymbolReference.cs]: okay to leak this language specific info to core layer?

File: src/Features/Core/Portable/AddImport/References/SymbolReference.cs [hunk 2]
```diff
@@ -87,19 +88,31 @@ public override int GetHashCode()
...
+                    // The name is going to change.  Make it clear in the description that this is
+                    // going to happen.
                     description = $"{this.SearchResult.DesiredName} - {description}";
+
+                    // if we were a fuzzy match, and we didn't have any glyph to show, then add the
+                    // namespace-glyph to this item. This helps indicate that not only are we fixing
+                    // the spelling of this name we are *also* adding a namespace.  This helps as we
+                    // have gotten feedback in the past that the 'using/import' addition was
+                    // unexpected.
+                    if (tags.IsDefaultOrEmpty)
```
Comment (heejaechang) [inline src/Features/Core/Portable/AddImport/References/SymbolReference.cs]: so, icon is not composite, but shows more important concept<?> I guess? all fix is adding import I guess? but if one add more than import like adding reference, then it shows that glyph?

**Comment by mavasani** [comment]
Merging the change, @CyrusNajmabadi, feel free to reply to Heejae's feedback or address it in a follow-up change.

File: src/Features/Core/Portable/AddImport/References/ProjectSymbolReference.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/AddImport/References/ProjectSymbolReference.cs]: in this case, yes.  it's not really doing anything more than saying: what is the project glyph for this project type.  given that this feature is really only about connection vb/c# projects, it's feels overly complex to have to introduce a language indirection.

**Comment by CyrusNajmabadi** [comment]
thanks @mavasani

**Pull request merged**